src/libsystemd/sd-device/test-sd-device.c

```diff
@@ -499,6 +499,37 @@ TEST(sd_device_enumerator_add_match_parent) {
         }
 }
 
+TEST(sd_device_enumerator_add_all_parents) {
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        int r;
+
+        assert_se(sd_device_enumerator_new(&e) >= 0);
+        assert_se(sd_device_enumerator_allow_uninitialized(e) >= 0);
+        /* See comments in TEST(sd_device_enumerator_devices). */
+        assert_se(sd_device_enumerator_add_match_subsystem(e, "bdi", false) >= 0);
+        assert_se(sd_device_enumerator_add_nomatch_sysname(e, "loop*") >= 0);
+        assert_se(sd_device_enumerator_add_match_subsystem(e, "net", false) >= 0);
+
+        /*enable all_parents*/
+        assert_se(sd_device_enumerator_add_all_parents(e) >= 0);
+        /*filter in only a subsystem*/
+        assert_se(sd_device_enumerator_add_match_subsystem(e, "block", true) >= 0);
+
+        int not_filtered_parent_count = 0;
+        FOREACH_DEVICE(e, dev) {
+                const char *subsystem;
+                r = sd_device_get_subsystem(dev, &subsystem);
+                if (r < 0) {
+                        assert_se(r == -ENOENT);
```

Comment:
Hmm, is the condition necessary? If so, then `test_matches()` needs to be adjusted, to make at least the parent device has valid subsystem.